### DISCUSSION TITLE
Customize cloud init best practices

### DISCUSSION THREAD

**Discussion by nilo85** [Joshua-Riek/ubuntu-rockchip#741] [Q&A]
Hello, I want to somehow easily setup a few near identical machines that will make a kubernetes cluster, but I am also a fan of configuration as code / automating things,
So I was curious if we can use cloud init for this.

howver, beeing a complete beginner to this topic, I know so far this distro relies on it to create user etc.

what I am struggling with is if there is an easy way to apply custom config to the prebuilt images? Or is best approach to fork and create my own variant of this distribution?

Anyone else done something similar? I know the raspberry pi imager offers customization support via cloud init but I guess it’s cause they structured their image in a specific way?

**Comment by nilo85** [comment]
It seems rpi images rely on partition 1 being FAT and then it writes "user-data" file there. 

running fdisk on one of the images it seems we only have a linux (ext4?) partition on the image so guess that makes custimization a bit harder.

```
fdisk ubuntu-22.04.3-preinstalled-server-arm64-orangepi-5-plus.img
Disk: ubuntu-22.04.3-preinstalled-server-arm64-orangepi-5-plus.img	geometry: 998/128/63 [8048640 sectors]
Signature: 0xAA55
         Starting       Ending
 #: id  cyl  hd sec -  cyl  hd sec [     start -       size]
------------------------------------------------------------------------
 1: EE    0   0   2 - 1023 255  63 [         1 -    8048639] <Unknown ID>
 2: 00    0   0   0 -    0   0   0 [         0 -          0] unused
 3: 00    0   0   0 -    0   0   0 [         0 -          0] unused
 4: 00    0   0   0 -    0   0   0 [         0 -          0] unused
 ```

rpi-imager seems to "mount" the fat FS in software and put the file there
https://github.com/raspberrypi/rpi-imager/blob/qml/src/downloadthread.cpp#L980

Guess we would need to restructure the disk layout of the image to get a similar feature?

EDIT: However it seems we do create a fat partition..? Maybe I built it with an too old version, or the tools I have on the mac is not super helpful =/
https://github.com/Joshua-Riek/ubuntu-rockchip/blob/main/scripts/build-image.sh#L96

**Reply by Joshua-Riek** [reply]
The partition layout is different for Ubuntu 24.04 images due to troubles with flash-kernel and how unfriendly it is when supporting multiple boards. However, there is a 4MB FAT partition in the server image where customizations can be made with the cloud-init config files.

**Reply by nilo85** [reply]
I guess I just need to rebuild the image then =) 3 weeks ago you created a new layout
https://github.com/Joshua-Riek/ubuntu-rockchip/commit/5c655e1d550c7925d6f9d30b00735519c7631c96

Is the cloud-init stored on this FAT partition?

**Reply by nilo85** [reply]
To answer my own follow up question, yes it seems to be the case =)
https://github.com/Joshua-Riek/ubuntu-rockchip/blob/main/scripts/build-image.sh#L147

**Reply by Joshua-Riek** [reply]
Sorry but I really can't be of more help right now, I'm so burnt out. I just want to get Ubuntu 24.04 out the door then I'm taking a long break. Ive honestly been hacking in code to the build system too because i dont care right now, so i really need a break.

**Reply by nilo85** [reply]
You are more than helpful as always! I think I have what I need to try stuff out and if I get something working maybe I can even contribute with some knowledge about it in the future =)

Take care, burning out is nasty, tried that last year and still recovering...

**Reply by tdesaules** [reply]
Hey @nilo85 did you find something about your initial need ? I have the same concern (before starting from scratch)

**Reply by nilo85** [reply]
@tdesaules Semi, what I ended up was to flash a server variant to an SD card, then I made bash scripts that flashed the eMMC, then mounts the fat32 and copies over custom cloud init config.

So I put the SD card in, boot the device, SSH to my known hostname for the "installer", then flash the desired configuration, power off and unmount and starts the device again.

Was considering if I wanted to do something more re-usable but at this point It works fine enough
